src/page/page_cache.rs

```diff
@@ -359,7 +395,211 @@ impl<M> PageCache<LockFreeLog, M>
             log_reservation.abort();
         } else {
             log_reservation.complete();
+            self.maybe_snapshot();
         }
         result
     }
+
+    fn maybe_snapshot(&self) {
+        let count = self.updates.fetch_add(1, SeqCst) + 1;
+        let should_snapshot = self.config().get_path().is_some() &&
+            count % self.config().get_snapshot_after_ops() == 0;
+        if !should_snapshot {
+            return;
+        }
+        self.log.flush();
+
+        let snapshot_opt = self.last_snapshot.take(SeqCst);
+        if snapshot_opt.is_none() {
+            // some other thread is snapshotting
+            warn!(
+                "snapshot skipped because previous attempt \
+                  appears not to have completed"
+            );
+            return;
+        }
```

Comment:
with the `AtomicOption`, we are enforcing the invariant that only one thread will ever be running a snapshot simultaneously. You need to `take()` it (which is atomic) and hold the old snapshot before you're allowed to create a new one.

This is more of a capacity-related invariant than anything, as it should be totally safe to create as many snapshots as you want, or delete them, etc... A snapshot is an immutable optimization over an immutable underlying log.